content/00.build/65.developer-reference/50.era-contracts/40.handling-l1-l2-ops.md

```diff
@@ -0,0 +1,185 @@
+and rely on L1 contracts to validate it, it may be a very powerful tool for a malicious operator.
+Note, that while the governance mechanism is generally trusted, we try to limit our trust for the operator as much as possible,
+since in the future anyone would be able to become an operator.
+
+Some time ago, we *used to* have a system where the upgrades could be done via L1→L2 transactions,
+i.e. the implementation of the `DiamondProxy` upgrade would
+[include](https://github.com/matter-labs/era-contracts/blob/f06a58360a2b8e7129f64413998767ac169d1efd/ethereum/contracts/zksync/upgrade-initializers/DIamondUpgradeInit2.sol#L27)
+a priority transaction (with `from` equal to for instance `FORCE_DEPLOYER`) with all the upgrade params.
+
+In the Boojum though having such logic would be dangerous and would allow for the following attack:
+
+- Let’s say that we have at least 1 priority operations in the priority queue.
+  This can be any operation, initiated by anyone.
+- The operator puts a malicious priority operation with an upgrade into the bootloader memory.
+  This operation was never included in the priority operations queue / and it is not an upgrade transaction.
+  However, as already mentioned above the bootloader has no idea what priority / upgrade transactions
+  are correct and so this transaction will be processed.
+
+The most important caveat of this malicious upgrade is that it may change
+implementation of the `Keccak256` precompile to return any values that the operator needs.
+
+- When the`priorityOperationsRollingHash` will be updated,
+  instead of the “correct” rolling hash of the priority transactions,
+  the one which would appear with the correct topmost priority operation is returned.
+  The operator can’t amend the behaviour of `numberOfPriorityTransactions`, but it won’t help much,
+  since the the `priorityOperationsRollingHash` will match on L1 on the execution step.
+
+That’s why the concept of the upgrade transaction is needed:
+this is the only transaction that can initiate transactions out of the kernel space and
+thus change bytecodes of system contracts.
+That’s why it must be the first one and that’s why
+[emit](https://github.com/code-423n4/2024-03-zksync/blob/main/code/system-contracts/bootloader/bootloader.yul#L609)
+its hash via a system L2→L1 log before actually processing it.
+
+### Why it doesn’t break on the previous version of the system
+
+This section is not required for current system understanding
+but for those willing to analyze the previous production system which wasn't susceptible to attack.
+
+Note that the hash of the transaction is calculated before the transaction is executed:
+[https://github.com/matter-labs/era-system-contracts/blob/3e954a629ad8e01616174bde2218241b360fda0a/bootloader/bootloader.yul#L1055](https://github.com/matter-labs/era-system-contracts/blob/3e954a629ad8e01616174bde2218241b360fda0a/bootloader/bootloader.yul#L1055)
```

Comment:
suggestion: We should replace the long url with some words so it's easier to read
```suggestion
[link to code](https://github.com/matter-labs/era-system-contracts/blob/3e954a629ad8e01616174bde2218241b360fda0a/bootloader/bootloader.yul#L1055)
```

It currently looks like this:
![image](https://github.com/matter-labs/zksync-docs/assets/1890113/deb3afa5-d12c-4aad-b0e8-1b7cfbbd0cca)